apps/src/lab2/lab2Redux.ts

```diff
@@ -286,28 +286,26 @@ export const isReadOnlyWorkspace = (state: RootState) => {
   const isStartMode = getAppOptionsEditBlocks() === START_SOURCES;
   const isEditingExemplarMode = getAppOptionsEditingExemplar();
   const isFrozen = !!state.lab.channel?.frozen;
-  const isPredictLevel =
-    state.lab.levelProperties?.predictSettings?.isPredictLevel || false;
-  let isReadonlyPredictLevel = isPredictLevel;
-  if (isPredictLevel) {
-    const isEditableAfterSubmit =
-      state.lab.levelProperties?.predictSettings?.codeEditableAfterSubmit ||
-      false;
-    const hasSubmittedPredictResponse = state.predictLevel.hasSubmittedResponse;
-    // If the predict level code is not editable after submit or the user has not submitted a response,
-    // the predict level is read only.
-    isReadonlyPredictLevel =
-      !isEditableAfterSubmit || !hasSubmittedPredictResponse;
-  }
+  const readonlyPredictLevel = isReadonlyPredictLevel(state);
   const hasSubmitted = getCurrentLevel(state)?.status === LevelStatus.submitted;
+  const isRunningAndReadonly =
+    state.lab2System.isRunning && shouldBeReadonlyWhileRunning(state);
```

Comment:
I wonder if there's a way to have the variable names mirror each other in a way that makes it clear which ones are "siblings" and which ones are "children" in terms of the level at which they get checked. It seems like `isReadOnly` is the parent here, and the other checks are all siblings to each other, so maybe something like...

```
export const isReadOnlyWorkspace = (state: RootState) => {
  // We are always in edit mode if we are in start or editing exemplar mode.
  // Both of these modes have no channel.
  const isInStartMode = getAppOptionsEditBlocks() === START_SOURCES;
  const isInEditingExemplarMode = getAppOptionsEditingExemplar();
  if (isInStartMode || isInEditingExemplarMode) {
    return false;
  }
  
  const isReadOnlyBecauseNotOwner = !state.lab.channel?.isOwner;
  const isReadOnlyBecauseFrozen = !!state.lab.channel?.frozen;
  const isReadOnlyBecausePredictLevel = isReadonlyPredictLevel(state);
  const isReadOnlyBecauseSubmitted = getCurrentLevel(state)?.status === LevelStatus.submitted;
  const isReadOnlyBecauseRunning =
    state.lab2System.isRunning && shouldBeReadonlyWhileRunning(state);

  // Otherwise, we are in read-only mode if we are not the owner of the channel,
  // the level is frozen, the level is a read-only predict level, the level has been submitted,
  // or this is a lab that should be read-only while running and the code is currently running.
  return (
    isReadOnlyBecauseNotOwner ||
    isReadOnlyBecauseFrozen ||
    isReadOnlyBecausePredictLevel ||
    isReadOnlyBecauseSubmitted ||
    isReadOnlyBecauseRunning
  );
};
```

I know that blows up the variable names, but maybe it's clearer?

Reply:
I rearranged the method to put the start/exemplar check before the other booleans are created, I has considered doing that previously.

I don't think we need the extra `isReadonlyBecause` on each variable, I feel like it's somewhat self-explanatory? I left `isOwner` as is, since in the return statement `!isOwner` is pretty readable. But that's a stylistic preference as to whether or not to have your booleans be always positive or not. I think I've heard a good argument here to keep all booleans positive (`isOwner` rather than `isNotOwner`) because you can end up with things like `!isNotOwner` which is hard to read. Although that wouldn't happen here and I broke that rule in my other PR, so I'm open to updating it if you think it's more readable.

Reply:
I have no strong feelings/agree about keeping booleans always positive (I didn't register that I'd flipped that one and how that would be perceived), but I did think about this:

> I'd be more inclined to use this style if it wasn't all wrapped up in a function. Since it's already isReadOnly... in the function name, I do agree that it's kinda redundant to have it on all of the variables inside of it. If these were several variables outside of a comparable function, I think the extra prefix would be beneficial.

I'm okay with leaving it as-is -- thanks for the discussion!